pgx-pg-sys/build.rs

```diff
@@ -392,23 +416,632 @@ fn impl_pg_node(
+            "Node" | "RangeTblEntry" | "List" => {
+                // use the special implementation defined above.
+                continue;
+            }
+            _ => {}
+        }
+        let mut traverse_elements: Vec<proc_macro2::TokenStream> = Vec::new();
+
+        for field in node_struct.struct_.fields.iter() {
+            let field_name = field.ident.as_ref().unwrap();
+            println!("Handling {} -> {} of type {:?}", struct_name, field_name, &field.ty);
+            // Some structures have array fields in them, rather than Lists. Handle those specially.
+            match array_fields
+                .get(&(struct_name.to_string().as_ref(), field_name.to_string().as_ref()))
+            {
+                Some(abi) => {
+                    match &abi.n {
+                        Some(n) => handle_length_bounded_array(
+                            &mut traverse_elements,
+                            n,
+                            field_name.to_string().as_ref(),
+                        ),
+                        _ => {}
+                    }
+                    continue;
+                }
+                _ => {}
+            }
+            // Some structures have fields that we need to handle specially because they're not populated, or are Lists
+            // that don't contain pointers.
+            match (struct_name.to_string().as_ref(), field_name.to_string().as_ref()) {
+                ("ModifyTableState", "mt_arowmarks")
+                | ("ModifyTableState", "mt_per_subplan_tupconv_maps")
+                | (_, "xpr") => continue,
+                _ => {}
+            }
+            match &field.ty {
+                Type::Array(a) => {
+                    if let Type::Ptr(ptr) = a.elem.as_ref() {
+                        if let Type::Path(p) = ptr.elem.as_ref() {
+                            if is_node_struct(
+                                &p.path.segments.first().unwrap().ident.to_string(),
+                                &node_types,
+                                &nodes,
+                            ) {
+                                does_walking = true;
+                                traverse_elements.push(quote!{
+                                    for item in self.#field_name.iter() {
+                                        if !item.is_null() {
+                                            walker_fn(item.clone() as *mut Node, context);
+                                            Node::traverse::<T>( unsafe { &mut *(*item as *mut Node) }, walker_fn, context);
+                                        }
+                                    }
+                                });
+                            }
+                        } else {
+                            ptr_problems.push(format!(
+                                "Unexpected type inside array:ptr {} -> field {}: {:?}",
+                                struct_name, field_name, ptr.elem
+                            ));
```

Comment:
I would kiiinda prefer it if this used a structured type and the formatting was done at panic time. But this is a fairly weak preference.

Reply:
Fair enough.